### ISSUE TITLE
Why doesn't MatterJS make more use of the prototype pattern?

### ISSUE THREAD

**Issue by Aralun** [liabru/matter-js#78]
Plop,

We use the MatterJS objects as follow:

    var myEngine = Matter.Engine.create(...);
    Matter.Engine.run(myEngine);

I personally love the prototype pattern, which would allow to write `myEngine.run()` instead of going through `Matter.Engine` again. Why doesn't MatterJS use it? I'm just curious about it.

**Comment by marlonicus** [comment]
Hey Aralun,

I can't speak for the author, but I used to love using a prototypal pattern too, especially coming from an ActionScript background. Then I read the following and now I much prefer the style that MatterJS is written in. These are slightly lengthy reads (at least by todays standards), but well worth it:

Two pillars of JavaScript ([Part 1](https://medium.com/javascript-scene/the-two-pillars-of-javascript-ee6f3281e7f3) / [Part 2](https://medium.com/javascript-scene/the-two-pillars-of-javascript-pt-2-functional-programming-a63aa53a41a4))

[You Don't Know JS](https://github.com/getify/You-Dont-Know-JS/) (In particular, the chapter on [this and object prototypes](https://github.com/getify/You-Dont-Know-JS/blob/master/this%20&%20object%20prototypes/ch6.md))

Hopefully they will give you a better understanding to the approach taken and the advantages that come with it!

**Comment by Aralun** [comment]
Well THANK YOU, I wanted to be productive today. GUESS NOT.

Thanks for the links. To me, it kinda seems like a pointless excess of code, so I'm definitely missing something. Coming back here tonight when I have read all that.

**Comment by Aralun** [comment]
15 days later, and after reading a good dozen of articles, I still don't understand why MatterJS is using this kind of code rather than using the Prototype chain. In my own limited vision of it, it's just inducing more code to write and not using one of JS most powerful feature. The articles that @marlonicus provided rather speak for the prototype pattern, not for the namespace use.

If this is not the right place to discuss it, please accept my apologies and close as resolved.

**Comment by liabru** [comment]
@marlonicus thanks for providing those links, they're a pretty good explanation - it's also interesting to see others advocating this, because until recently I'd not seen many other people doing it in JS!

Matter.js is largely written in functional style code rather than OOP. This is mostly because it's a style of coding I prefer in general, rather than any big decision about its advantages for Matter.js in particular. It was a bit of an experiment to use it on a larger project like this, but I like how it turned out.

I like this style because it makes my code more modular and composable both for functions and objects. It also helps promote loose coupling, which is easier to reason about and test. Although they don't all fully apply to JS, I really like the ideas of pure functions, non-mutability, minimal side effects and avoiding state etc.

The way I see it, all objects in Matter.js are just data models describing the state of a world at a particular instance in time. You can therefore describe a complete Matter.js world purely in JSON and then run it as-is with out any further interpretation (actually I'm lying, technically you need to handle circular references due to limitations of JSON, there are [libs](https://github.com/skeeto/resurrect-js) that can do this).

The functions in Matter.js are grouped by namespaced modules, but that doesn't mean that they are actually tied to any particular type of object, it's just a convenient logical organisation. 

I like to leverage duck typing where appropriate. Given that the inputs satisfy the dependencies of these functions, they will happily do their work on what ever types of object you supply. A good example in Matter.js of this at work is the [Events](https://github.com/liabru/matter-js/blob/master/src/core/Events.js) module.

That events module will not only work on any object in Matter.js, it will work on any object in JavaScript. You can lift that module as-is and use it standalone in any JS project you like with literally no changes to the code in that module or changes in your own code, it should just work (well, assuming there's no naming clashes).

The same goes for many other modules, including the [Vector](https://github.com/liabru/matter-js/blob/master/src/geometry/Vector.js) module. Again, this will work on any objects that define the properties `x` and `y`, it doesn't matter what prototype they inherit from.

Note that the `object.method()` style syntax itself isn't incompatible with the above concepts (as you can simply consider this as and example of [partial application](https://en.wikipedia.org/wiki/Partial_application)), but it would be tricky to implement this in a clean way into the current code base.

Basically, it's worth looking into functional programming if you haven't already!

**Comment by Aralun** [comment]
Oh, I see know. Perfect timing because I also happen, thanks to @marlonicus, to delve into this weird world of functional programming... I enjoy it a lot. I actually am applying it right now in a couple side-projects.

I didn't even think about MatterJS this way. It makes a lot of sense. Thanks for the insights!

**Comment by liabru** [comment]
No problem, learning about FP really changes the way you think about programming, it's definitely worth it!

**Comment by MegaUITeam** [comment]
@liabru 

1) winnings 0
2) the excess code is
3) your problem is solved by mixins.
